src/main/java/org/apache/hadoop/mapred/MesosExecutor.java

```diff
@@ -55,52 +67,79 @@ public void disconnected(ExecutorDriver driver) {
   public void launchTask(final ExecutorDriver driver, final TaskInfo task) {
     LOG.info("Launching task : " + task.getTaskId().getValue());
 
-    // Get configuration from task data (prepared by the JobTracker).
-    JobConf conf = configure(task);
+    synchronized(this) {
 
-    // NOTE: We need to manually set the context class loader here because,
-    // the TaskTracker is unable to find LoginModule class otherwise.
-    Thread.currentThread().setContextClassLoader(
-        TaskTracker.class.getClassLoader());
+      // Keep track of all the IDs we've been asked to monitor
+      if (task.getTaskId().getValue().endsWith("_Map")) {
+        mapTaskId = task.getTaskId();
+      } else if (task.getTaskId().getValue().endsWith("_Reduce")) {
+        reduceTaskId = task.getTaskId();
+      } else {
+        driver.sendStatusUpdate(TaskStatus.newBuilder()
+          .setTaskId(task.getTaskId())
+          .setState(TaskState.TASK_LOST).build());
 
-    try {
-      taskTracker = new TaskTracker(conf);
-    } catch (IOException | InterruptedException e) {
-      LOG.fatal("Failed to start TaskTracker", e);
-      System.exit(1);
-    }
+        return;
+      }
 
-    // Spin up a TaskTracker in a new thread.
-    new Thread("TaskTracker Run Thread") {
-      @Override
-      public void run() {
-        try {
-          taskTracker.run();
+      if (taskTracker == null) {
+        // Get configuration from task data (prepared by the JobTracker).
+        JobConf conf = configure(task);
 
-          // Send a TASK_FINISHED status update.
-          // We do this here because we want to send it in a separate thread
-          // than was used to call killTask().
-          driver.sendStatusUpdate(TaskStatus.newBuilder()
-              .setTaskId(task.getTaskId())
-              .setState(TaskState.TASK_FINISHED)
-              .build());
+        // NOTE: We need to manually set the context class loader here because,
+        // the TaskTracker is unable to find LoginModule class otherwise.
+        Thread.currentThread().setContextClassLoader(
+            TaskTracker.class.getClassLoader());
 
-          // Give some time for the update to reach the slave.
-          try {
-            Thread.sleep(2000);
-          } catch (InterruptedException e) {
-            LOG.error("Failed to sleep TaskTracker thread", e);
-          }
-
-          // Stop the executor.
-          driver.stop();
-        } catch (Throwable t) {
-          LOG.fatal("Caught exception, committing suicide.", t);
-          driver.stop();
+        try {
+          taskTracker = new TaskTracker(conf);
+        } catch (IOException | InterruptedException e) {
+          LOG.fatal("Failed to start TaskTracker", e);
           System.exit(1);
         }
+
+        // Spin up a TaskTracker in a new thread.
+        new Thread("TaskTracker Run Thread") {
+          @Override
+          public void run() {
+            try {
+              taskTracker.run();
```

Comment:
I guess it was like this before, but the code was rearranged. I'll take it that it is safe.